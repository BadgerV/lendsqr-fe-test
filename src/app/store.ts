import { configureStore } from "@reduxjs/toolkit";
import UserInfoSlice from "./UserInfoSlice/UserInfoSlice";

export const store = configureStore({
  reducer: {
    UserInfoSlice: UserInfoSlice,
  },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
