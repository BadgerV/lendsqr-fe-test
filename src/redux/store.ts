import { configureStore, combineReducers } from "@reduxjs/toolkit";
import UserInfoSlice from "./UserInfoSlice/UserInfoSlice";

export const store = configureStore({
  reducer: {
    UserInfoSlice: UserInfoSlice,
  },
});

// Create the root reducer separately so we can extract the RootState type
const rootReducer = combineReducers({
  UserInfoSlice: UserInfoSlice,
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
export type AppStore = ReturnType<typeof setupStore>;

export const setupStore = (preloadedState?: Partial<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};
