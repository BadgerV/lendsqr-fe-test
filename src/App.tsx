// Importing pages for routing
import Dashboard from "./pages/Dashboard/Dashboard";
import User from "./pages/User/User";
import Login from "./pages/Login/Login";

// Importing React hooks and components
import { useEffect } from "react";
import { Route, Routes, Navigate } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { AppDispatch } from "./redux/store";
import {
  fetchUserData,
  getPage,
  setUsersInfo,
} from "./redux/UserInfoSlice/UserInfoSlice";

// Importing user type from the types file
import { User as UserData } from "./types";
import { RootState } from "./redux/store";

import { useNavigate } from "react-router-dom";

// Define the main App component
const App: React.FC = () => {
  // Initialize the Redux dispatch function
  const dispatch = useDispatch<AppDispatch>();

  const navigate = useNavigate();

  const loggedInUser = useSelector(
    (state: RootState) => state.UserInfoSlice.loggedInUser
  );

  // Retrieve stored user data from local storage
  const userDataString = localStorage.getItem("userData");

  // Parse the stored user data, or default to an empty object if null
  const storedUserData: UserData | null = userDataString
    ? JSON.parse(userDataString)
    : null;

  // useEffect to fetch user data on component mount
  useEffect(() => {
    // Check if user data is not stored, then fetch and set the data
    if (!storedUserData) {
      dispatch(fetchUserData());
      dispatch(getPage());
    } else {
      // Set user data from local storage if available
      dispatch(setUsersInfo(storedUserData));
    }
  }, []);

  useEffect(() => {
    if (!loggedInUser) {
      navigate("/login");
    }
  }, [navigate, loggedInUser]);

  // Render the main application structure
  return (
    <div className="app" data-testid="app-container">
      {/* Set up the React Router for navigation */}
      <Routes>
        {/* Redirect from "/" to "/dashboard" */}
        <Route path="/" element={<Navigate to="/dashboard" />} />
        {/* Define routes for different pages */}
        <Route path="/dashboard" element={<Dashboard />} />
        <Route path="/login" element={<Login />} />
        <Route path="/user" element={<User />} />
      </Routes>
    </div>
  );
};

// Export the main App component
export default App;
