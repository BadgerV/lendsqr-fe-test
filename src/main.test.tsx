// testing main.test

import App from "./App";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { setupStore } from "./redux/store";

// rendering main App
describe("App component", () => {
  //mock store
  const store = setupStore();
  it("renders without crashing", () => {
    const { container } = render(
      <BrowserRouter>
        <Provider store={store}>
          <App />
        </Provider>
      </BrowserRouter>
    );
    // Check if the rendered component matches the snapshot
    expect(container).toMatchSnapshot();
  });
});
