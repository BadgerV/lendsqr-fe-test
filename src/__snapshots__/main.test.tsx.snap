// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > renders without crashing 1`] = `
<div>
  <div
    class="app"
    data-testid="app-container"
  >
    <div
      class="login"
    >
      <div
        class="login-content"
      >
        <div
          class="login-content__left"
        >
          <div
            class="login-content__left-logo-container"
          >
            <img
              alt="logo"
              src="/assets/logo.png"
            />
            <span>
              lendsqr
            </span>
          </div>
          <img
            alt="Login Page Left Image"
            class="login-image"
            src="/assets/pablo-sign-in.png"
          />
        </div>
        <div
          class="login-content__right"
        >
          <span
            class="login-content__right-welcome-text"
          >
            Welcome!
          </span>
          <span
            class="login-content__right-details-text"
          >
            Enter details to login.
          </span>
          <form
            class="login-content__right-form"
          >
            <div
              class="login-content__right-form-input-container"
            >
              <input
                placeholder="Username"
                type="text"
                value=""
              />
            </div>
            <div
              class="login-content__right-form-input-container"
            >
              <input
                placeholder="Password"
                type="password"
                value=""
              />
              <span
                class="login-content__right-form-input-container-show-text"
              >
                SHOW
              </span>
            </div>
            <span
              class="login-content__right-forgot-password"
            >
              FORGOT PASSWORD?
            </span>
            <button
              class="login-content__right-submit-btn"
              disabled=""
            >
              LOG IN
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
